stdlib/public/stubs/LibcShims.cpp

```diff
@@ -308,3 +318,72 @@ swift::_stdlib_cxx11_mt19937_uniform(__swift_uint32_t upper_bound) {
   std::uniform_int_distribution<__swift_uint32_t> RandomUniform(0, upper_bound);
   return RandomUniform(getGlobalMT19937());
 }
+
+// _stdlib_random
+//
+// Should the implementation of this function add a new platform/change for a
+// platform, make sure to also update the stdlib documentation regarding
+// platform implementation of this function.
+// This can be found at: stdlib/public/core/Random.swift
+
+#if defined(_WIN32) && !defined(__CYGWIN__)
+#error TODO: Test _stdlib_random on Windows
+
+SWIFT_RUNTIME_STDLIB_INTERNAL
+void swift::_stdlib_random(void *buf, __swift_size_t nbytes) {
+  NTSTATUS status = BCryptGenRandom(nullptr,
+                                    static_cast<PUCHAR>(buf),
+                                    static_cast<ULONG>(nbytes),
+                                    BCRYPT_USE_SYSTEM_PREFERRED_RNG);
+  if (!BCRYPT_SUCCESS(status)) {
+    fatalError(0, "Fatal error: 0x%.8X in '%s'\n", status, __func__);
+  }
+}
+
+#else
+
+#undef  WHILE_EINTR
+#define WHILE_EINTR(expression) ({                                             \
+  decltype(expression) result = -1;                                            \
+  do { result = (expression); } while (result == -1 && errno == EINTR);        \
+  result;                                                                      \
+})
+
+SWIFT_RUNTIME_STDLIB_INTERNAL
+void swift::_stdlib_random(void *buf, __swift_size_t nbytes) {
+  while (nbytes > 0) {
+    __swift_ssize_t actual_nbytes = -1;
+#if defined(GRND_RANDOM)
+    static const bool getrandom_available =
+      !(getrandom(nullptr, 0, 0) == -1 && errno == ENOSYS);
+    if (getrandom_available) {
+      actual_nbytes = WHILE_EINTR(getrandom(buf, nbytes, 0));
```

Comment:
Is a blocking read from the OS's entropy source the right behavior here?  For applications like server side Swift we can pretty well expect a fast response from the OS because it'll have initialized its entropy sources by the time a Swift process is up.  But for embedded platforms this call might take as long as minutes if the OS deems there's insufficient entropy.

There's pros and cons either way, I would just like to see a rationale for this somewhere for reference.

Reply:
The version of `getrandom` we're using here only blocks on OS startup where `/dev/urandom` may not have been initialized yet. After `/dev/urandom` is initialized, `getrandom` will always call into it, which makes this call non-blocking 99.99% of the time. The [proposal](https://github.com/apple/swift-evolution/blob/master/proposals/0202-random-unification.md#shouldnt-this-fallback-on-something-more-secure-at-times-of-low-entropy) explains this and shares that Python's implementation uses `getrandom()` as well, but falls back on `/dev/urandom` if `getrandom()` deems it needed to block on OS startup. We could do something similar.

Reply:
> The proposal explains this and shares that Python's implementation uses getrandom() as well, but falls back on /dev/urandom if getrandom() deems it needed to block on OS startup. We could do something similar.

We could, but doesn't Python 3.6 do what you've got here now?

Reply:
Ah you're right, 3.6 does what we have here and 3.5 fell back on reading `/dev/urandom`.